### DISCUSSION TITLE
create_mesh issue on a two materials foil

### DISCUSSION THREAD

**Discussion by jbocreative** [robbievanleeuwen/section-properties#428] [Q&A] [answered]
Hi all
I am trying to mesh these of two geometries with different materials.
![Figure_1](https://github.com/robbievanleeuwen/section-properties/assets/167868870/4f58c372-eb41-4f40-96ff-d5e5680aed06)

The mesh process starts but never ends

Both geometries have been meshed individually successfully.
I tried also to create construction points before merging the geometries.

Any clues ?

```
from sectionproperties.pre import Geometry

## structure

points=[(243.2,59.7),(548.5,49.3),(547.7,26.0),(483.4,28.2),(476.3,21.6),(475.6,0.0),(460.4,0.0),(461.4,29.0),
        (272.8,35.4),(271.6,0.0),(261.2,0.0),(261.9,20.9),(247.6,36.2),(242.4,36.4)]

facets = []
for i in range(len(points) - 1):
    facets.append((i, i + 1))
facets.append((len(points) - 1, 0))  # connect last to first

control_points = [(points[0][0] + 1,
                   points[0][1] -1)]

struct_geom = Geometry.from_points(
    points=points,
    facets=facets,
    control_points=control_points,
)
# struct_geom.plot_geometry()

## Fairing

points=[(5.4,0.0),(6.4,5.0),(9.2,9.8),(12.0,12.9),(16.0,16.4),(21.2,20.0),(27.9,23.8),(36.7,27.8),(41.8,29.9),
        (47.8,32.1),(54.6,34.3),(62.4,36.6),(71.2,39.0),(81.0,41.3),(92.0,43.7),(103.9,45.9),(116.8,48.1),(130.4,50.1),
        (144.6,51.8),(159.3,53.4),(174.4,54.8),(189.8,56.1),(205.5,57.1),(221.3,57.9),(237.4,58.6),(253.6,59.0),
        (269.9,59.4),(286.4,59.6),(302.9,59.6),(319.6,59.5),(336.3,59.3),(353.1,59.0),(369.9,58.5),(386.9,58.0),
        (403.9,57.3),(420.9,56.5),(438.0,55.7),(455.1,54.8),(472.2,53.7),(489.4,52.6),(506.6,51.5),(523.9,50.2),
        (541.2,48.9),(558.5,47.5),(575.8,46.1),(593.1,44.6),(610.5,43.1),(627.9,41.5),(645.3,39.8),(662.6,38.1),
        (680.0,36.3),(699.4,34.3),(700.0,39.7),(680.6,41.7),(663.2,43.5),(645.8,45.2),(628.4,46.8),(611.0,48.4),
        (593.6,50.0),(576.3,51.5),(558.9,52.9),(541.6,54.3),(524.3,55.6),(507.0,56.9),(489.8,58.0),(472.6,59.1),
        (455.4,60.2),(438.2,61.1),(421.1,61.9),(404.1,62.7),(387.1,63.4),(370.1,63.9),(353.2,64.4),(336.4,64.7),
        (319.6,64.9),(302.9,65.0),(286.3,65.0),(269.8,64.8),(253.4,64.5),(237.2,64.0),(221.1,63.3),(205.1,62.5),
        (189.4,61.4),(173.9,60.2),(158.7,58.8),(143.9,57.2),(129.6,55.4),(115.9,53.4),(103.0,51.3),(90.9,49.0),
        (79.8,46.6),(69.8,44.2),(60.9,41.8),(53.0,39.5),(46.0,37.2),(39.9,34.9),(34.5,32.8),(25.5,28.6),(18.4,24.6),
        (12.7,20.7),(8.2,16.8),(4.7,12.8),(1.3,6.9),(0.0,0.0)]

facets = []
for i in range(len(points) - 1):
    facets.append((i, i + 1))
facets.append((len(points) - 1, 0))  # connect last to first

control_points = [(1,1)]

fair_geom = Geometry.from_points(
    points=points,
    facets=facets,
    control_points=control_points,
)

## Union geometries

temp_geom = (fair_geom - struct_geom) | fair_geom
geom = (struct_geom - fair_geom) + temp_geom

# geom.plot_geometry()

# geom.create_mesh(mesh_sizes=6)
```

**Comment by robbievanleeuwen** [comment, marked as answer]
Hi @jbocreative I had a play around with your geometry and could not get it to work, I suspect the issue is the extremely sharp angle between the fairing and the structural geometry. I believe the meshing algorithm first meshes this area and then eats away the holes later, so I think it is probably having an issue in this area. I also tried changing the `min_angle` parameter of `create_mesh()` to no avail.

<img width="760" alt="image" src="https://github.com/robbievanleeuwen/section-properties/assets/18841910/e9cd6259-fb84-4776-b55e-8de6c4978bab">

If you are only interested in geometric and plastic properties, the mesh converges by setting `coarse=True`, however you have no real control over the mesh density and may get inaccurate results for warping analyses (which is mesh dependent unlike the other two analyses).

![Screenshot 2024-05-31 at 12 19 55 pm](https://github.com/robbievanleeuwen/section-properties/assets/18841910/ac50f857-8b54-4ccb-9d96-f48caf82f94c)

You could "artificially" improve the mesh by introducing more nodes in your geometry, which would be forced to be within the mesh but this would be a bit more pre-processing work for you. Probably the best solution would be to change your geometry so that you don't have such a sharp junction, e.g. square it off - I haven't tried this but suspect it would work. The results should not be vastly different.

Sorry I don't have a better answer! Best of luck!

**Comment by jbocreative** [comment]
Hi
Thanks for trying.
I made several attempts myself.
The only thing that would eventually work is splitting and reassembling.
The issue is that the following does note create all the nodes it should
`temp_geom = (fair_geom - struct_geom) | fair_geom`
Anyway I ended up with analytically placing intersection nodes on both geometries.
then it is possible to compound them and mesh the whole.
Globally I have a poor experience with operations on geometries within sectionproperties.

**Reply by robbievanleeuwen** [reply]
I'm not sure this is the issue. I get perfect geometry without any node or facet issues by using:

```python
geom = fair_geom - struct_geom  + struct_geom
geom.plot_geometry(labels=("points","facets"))
```

I wouldn't use the union here as I don't think it's necessary or appropriate. Glad you got it working eventually 👍
